File: go/pull-with-push-override/config/cfg-src.go
/*
 *  Copyright © 2022 Paulo Villela. All rights reserved.
 *  Use of this source code is governed by the MIT license
 *  that can be found in the LICENSE file.
 */

package config

type CfgSrc[T any] func() T

func nilCfgSrc[T any]() T {
	panic("Module used before being initialized")
}

func MakeCfgSrc[T any](adapter func(src AppCfgInfo) T) CfgSrc[T] {
	if adapter == nil {
		return nilCfgSrc[T]
	}
	return func() T { return adapter(GetAppConfiguration()) }
}

File: ts/src/pull-with-push-override/config/cfg-src.ts
import { ConfigurationError } from "../../arch/util/configuration-error";
import { AppCfgInfo, getAppConfiguration } from "./app-cfg";
import { makeMutableFunctionArity0, MutableFunctionArity0 } from "../../arch/util/mutable-function-arity-0";

export type CfgSrc<T> = MutableFunctionArity0<T>

function nilAdapter<T>(info: AppCfgInfo): T {
	throw new ConfigurationError("Module used before being initialized")
}

export function makeCfgSrc<T>(adapter: ((info: AppCfgInfo) => T) | null): CfgSrc<T> {
	const adapter1 = adapter || nilAdapter
	const initialSrc = () => adapter1(getAppConfiguration())
	return makeMutableFunctionArity0(initialSrc)
}

export function identity<T>(t: T): T {
	return t
}
